Add LQR latency compensation

pietroglyph opened the pull request:
This adds an extra measurement delay input box in the data analyzer, along with appropriate defaults for the various motor controller presets. Note that this changes the behavior of the loop type dropdown so that it now applies the presets over whatever the user has when the dropdown is changed; this is to make it so that the measurement delay preset is changed when the user changes the loop type automatically.

Closes #124.

calcmogul commented on frc_characterization/arm_characterization/data_analyzer.py:
```diff
@@ -965,6 +1000,22 @@ def _calcGains(kv, ka, qp, qv, effort, period):
+        # can be quite large as a result of filtering for some motor
+        # controller and sensor combinations.
+        #
+        # We use equation (11) from
+        # https://www.textroad.com/pdf/JBASR/J.%20Basic.%20Appl.%20Sci.%20Res.,%202(3)2607-2618,%202012.pdf
+        # (Note that equation (11) has a typo of P^-1 when they meant R^-1.)
```
This isn't actually equation (11). That paper only provides a continuous version of the formula, while we're using a discrete version. I put a derivation for it in E.4.2 "Time delay compensation, discrete case" in https://file.tavsys.net/control/controls-engineering-in-frc.pdf.

pietroglyph commented:
@calcmogul I went ahead and updated the comments to link to your book which has the discrete version. I also added the explanation on why the gain will be overly conservative. Let me know what you think.

pietroglyph commented:
@Oblarg Sorry to bother you, but could you take a look at merging this PR?

calcmogul commented:
That fact should be documented in a comment, so people can trace the origin of the magic numbers.

pietroglyph commented:
Alright, I think I addressed your concerns @Oblarg. A couple of notes:

 1. Tyler and I think that the delay for a 40 ms window should be (40-1)/2, not 40/2. I added a comment in the code that has a proof for this.
 2. I already had a delay for the brushed SPARK MAXes in velocity mode. You might have missed it because only some of the characteriztion tools have nonzero filtering delays (because e.g. the arm and elevator are position mode only.) I added a comment to clear up this confusion.
 3. I couldn't find any information on velocity filtering for the Talon SRX, so I set their measurement delay to zero and added a comment.

Oblarg commented on frc_characterization/drive_characterization/data_analyzer.py:
```diff
@@ -563,37 +580,44 @@ def setMeasurementDelay(delay):
                 STATE.period.set(0.001),
                 STATE.controller_time_normalized.set(True),
                 STATE.controller_type.set("Talon"),
+                # No known filtering/delay (other than CAN latency)
```
The Talon SRX default latency is the same as the Spark MAX brushed default latency, iirc, so I don't think this is correct.

pietroglyph replied:
You got me to reread the CTRE docs, and I now see that I misread them as only applying to the Falon FXes. However, [the docs](https://phoenix-documentation.readthedocs.io/en/latest/ch14_MCSensor.html#changing-velocity-measurement-parameters) seem to indicate that the Talon SRX delay is the same as the Talon FX delay (i.e. both are 131.5 ms.) I've pushed a change to fix this.
